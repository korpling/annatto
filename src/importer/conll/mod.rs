use std::{
    io::Read,
    path::{Path, PathBuf},
};

use encoding_rs_io::DecodeReaderBytes;
use graphannis::{
    model::AnnotationComponentType,
    update::{GraphUpdate, UpdateEvent},
};
use graphannis_core::graph::ANNIS_NS;
use itertools::Itertools;
use pest::{
    iterators::{Pair, Pairs},
    Parser,
};
use pest_derive::Parser;

use crate::{
    error::AnnattoError, util::graphupdate::path_structure, workflow::StatusSender, Module,
};

use super::Importer;

pub const MODULE_NAME: &str = "import_conll";

pub struct ImportCoNLL {}

impl Default for ImportCoNLL {
    fn default() -> Self {
        Self {}
    }
}

impl Module for ImportCoNLL {
    fn module_name(&self) -> &str {
        MODULE_NAME
    }
}

impl Importer for ImportCoNLL {
    fn import_corpus(
        &self,
        input_path: &std::path::Path,
        properties: &std::collections::BTreeMap<String, String>,
        tx: Option<crate::workflow::StatusSender>,
    ) -> Result<graphannis::update::GraphUpdate, Box<dyn std::error::Error>> {
        let mut update = GraphUpdate::default();
        let paths_and_node_names = path_structure(&mut update, input_path, &["conll", "conllu"])?;
        for (pathbuf, doc_node_name) in paths_and_node_names {
            self.import_document(&mut update, pathbuf.as_path(), doc_node_name, &tx)?;
        }
        Ok(update)
    }
}

impl ToString for Rule {
    fn to_string(&self) -> String {
        match self {
            Rule::lemma => "lemma".to_string(),
            Rule::upos => "upos".to_string(),
            Rule::xpos => "xpos".to_string(),
            Rule::deprel => "deprel".to_string(),
            Rule::enhanced_rel => "rel".to_string(),
            _ => "".to_string(),
        }
    }
}

impl ImportCoNLL {
    fn import_document(
        &self,
        update: &mut GraphUpdate,
        document_path: &Path,
        document_node_name: String,
        tx: &Option<StatusSender>,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let f = std::fs::File::open(document_path)?;
        let mut decoder = DecodeReaderBytes::new(f);
        let mut file_content = String::new();
        decoder.read_to_string(&mut file_content)?; // TODO this needs to be buffered. UD Files can be very large
        let conllu: Pairs<Rule> = CoNLLUParser::parse(Rule::conll, &file_content)?;
        self.map_document(update, document_node_name, conllu, tx)?;
        Ok(())
    }

    fn map_document(
        &self,
        update: &mut GraphUpdate,
        document_node_name: String,
        mut conllu: Pairs<Rule>,
        tx: &Option<StatusSender>,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let mut token_names = Vec::new();
        if let Some(pair) = conllu.next() {
            if pair.as_rule() == Rule::conll {
                for sentence in pair.into_inner() {
                    // iterate over sentences
                    if sentence.as_rule() == Rule::sentence {
                        token_names.extend(self.map_sentence(
                            update,
                            document_node_name.as_str(),
                            sentence,
                            tx,
                        )?);
                    }
                }
            }
        } else {
            if let Some(sender) = tx {
                let msg = format!("Could not parse file as conllu: {document_node_name}");
                let err = AnnattoError::Import {
                    reason: msg,
                    importer: self.module_name().to_string(),
                    path: PathBuf::from(document_node_name),
                };
                sender.send(crate::workflow::StatusMessage::Failed(err))?;
            }
        }
        for (source, target) in token_names.iter().tuple_windows() {
            update.add_event(UpdateEvent::AddEdge {
                source_node: source.to_string(),
                target_node: target.to_string(),
                layer: ANNIS_NS.to_string(),
                component_type: AnnotationComponentType::Ordering.to_string(),
                component_name: "".to_string(),
            })?;
        }
        Ok(())
    }

    fn map_sentence(
        &self,
        update: &mut GraphUpdate,
        document_node_name: &str,
        sentence: Pair<Rule>,
        tx: &Option<StatusSender>,
    ) -> Result<Vec<String>, Box<dyn std::error::Error>> {
        let mut token_names = Vec::new();
        let mut dependencies = Vec::new();
        let mut s_annos = Vec::new();
        let (l, c) = sentence.line_col();
        for member in sentence.into_inner() {
            match member.as_rule() {
                Rule::token => {
                    let (tok_name, dep) = self.map_token(update, document_node_name, member, tx)?;
                    token_names.push(tok_name);
                    dependencies.push(dep);
                }
                Rule::s_anno => {
                    let mut name = None;
                    let mut value = None;
                    for name_or_s_value in member.into_inner() {
                        match name_or_s_value.as_rule() {
                            Rule::name => {
                                name = Some(name_or_s_value.as_str().trim().to_string());
                            }
                            Rule::s_value => {
                                value = Some(name_or_s_value.as_str().trim().to_string())
                            }
                            _ => {}
                        }
                        if name.is_some() && value.is_some() {
                            s_annos.push((name.unwrap(), value.unwrap()));
                            break;
                        }
                    }
                }
                _ => {}
            }
        }
        if !token_names.is_empty() {
            let node_name = format!("{document_node_name}#{l}_{c}");
            update.add_event(UpdateEvent::AddNode {
                node_name: node_name.to_string(),
                node_type: "node".to_string(),
            })?;
            for (anno_name, anno_value) in s_annos {
                update.add_event(UpdateEvent::AddNodeLabel {
                    node_name: node_name.to_string(),
                    anno_ns: "".to_string(),
                    anno_name: anno_name.to_string(),
                    anno_value: anno_value.to_string(),
                })?;
            }
            for token_name in &token_names {
                update.add_event(UpdateEvent::AddEdge {
                    source_node: node_name.to_string(),
                    target_node: token_name.to_string(),
                    layer: ANNIS_NS.to_string(),
                    component_type: AnnotationComponentType::Coverage.to_string(),
                    component_name: "".to_string(),
                })?;
            }
            for (node_index, dependency) in dependencies.into_iter().enumerate() {
                let (head_id, deprel) = dependency.unwrap();
                let head_index = head_id - 1;
                if head_index < token_names.len() {
                    let source_node_name = token_names.get(head_index).unwrap();
                    let target_node_name = token_names.get(node_index).unwrap();
                    update.add_event(UpdateEvent::AddEdge {
                        source_node: source_node_name.to_string(),
                        target_node: target_node_name.to_string(),
                        layer: "".to_string(),
                        component_type: AnnotationComponentType::Pointing.to_string(),
                        component_name: "dep".to_string(),
                    })?;
                    if let Some(deprel_value) = deprel {
                        update.add_event(UpdateEvent::AddEdgeLabel {
                            source_node: source_node_name.to_string(),
                            target_node: target_node_name.to_string(),
                            layer: "".to_string(),
                            component_type: AnnotationComponentType::Pointing.to_string(),
                            component_name: "dep".to_string(),
                            anno_ns: "".to_string(),
                            anno_name: "deprel".to_string(),
                            anno_value: deprel_value.to_string(),
                        })?;
                    }
                }
            }
        }
        Ok(token_names)
    }

    fn map_token(
        &self,
        update: &mut GraphUpdate,
        document_node_name: &str,
        mut token: Pair<Rule>,
        tx: &Option<StatusSender>,
    ) -> Result<(String, Option<(usize, Option<String>)>), Box<dyn std::error::Error>> {
        let (l, c) = token.line_col();
        let node_name = format!("{document_node_name}#{l}_{c}");
        update.add_event(UpdateEvent::AddNode {
            node_name: node_name.to_string(),
            node_type: "node".to_string(),
        })?;
        update.add_event(UpdateEvent::AddNodeLabel {
            node_name: node_name.to_string(),
            anno_ns: ANNIS_NS.to_string(),
            anno_name: "layer".to_string(),
            anno_value: "default_layer".to_string(),
        })?;
        let mut head_id = None;
        let mut deprel = None;
        for member in token.into_inner() {
            let rule = member.as_rule();
            match rule {
                Rule::id => {}
                Rule::form => {
                    update.add_event(UpdateEvent::AddNodeLabel {
                        node_name: node_name.to_string(),
                        anno_ns: ANNIS_NS.to_string(),
                        anno_name: "tok.".to_string(),
                        anno_value: member.as_str().to_string(),
                    })?;
                }
                Rule::lemma | Rule::upos | Rule::xpos => {
                    let anno_name = rule.to_string();
                    update.add_event(UpdateEvent::AddNodeLabel {
                        node_name: node_name.to_string(),
                        anno_ns: ANNIS_NS.to_string(),
                        anno_name: anno_name.to_string(),
                        anno_value: member.as_str().to_string(),
                    })?;
                }
                Rule::features | Rule::misc => {
                    for feature in member.into_inner() {
                        if feature.as_rule() == Rule::feature {
                            let mut anno_name = None;
                            let mut anno_value = None;
                            for name_or_value in feature.into_inner() {
                                let feature_rule = name_or_value.as_rule();
                                if feature_rule == Rule::name {
                                    anno_name = Some(name_or_value.as_str().to_string());
                                } else if feature_rule == Rule::value {
                                    anno_value = Some(name_or_value.as_str().to_string());
                                }
                                if anno_name.is_some() && anno_value.is_some() {
                                    update.add_event(UpdateEvent::AddNodeLabel {
                                        node_name: node_name.to_string(),
                                        anno_ns: "".to_string(),
                                        anno_name: anno_name.unwrap().trim().to_string(),
                                        anno_value: anno_value.unwrap().trim().to_string(),
                                    })?;
                                    break;
                                }
                            }
                        } else if feature.as_rule() == Rule::no_value {
                            break;
                        }
                    }
                }
                Rule::head => {
                    for id_or_else in member.into_inner() {
                        match id_or_else.as_rule() {
                            Rule::id => {
                                head_id = Some(id_or_else.as_str().trim().parse::<usize>()?);
                                break;
                            }
                            _ => {}
                        }
                    }
                }
                Rule::deprel => {
                    deprel = Some(member.as_str().trim().to_string());
                }
                Rule::enhanced_deps => {}
                _ => {}
            }
        }
        let dependency = match head_id {
            None => None,
            Some(v) => Some((v, deprel)),
        };
        Ok((node_name, dependency))
    }
}

#[derive(Parser)]
#[grammar = "importer/conll/conllu.pest"]
struct CoNLLUParser;
