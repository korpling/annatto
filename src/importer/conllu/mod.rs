use std::{
    collections::BTreeMap,
    io::Read,
    path::{Path, PathBuf},
};

use encoding_rs_io::DecodeReaderBytes;
use graphannis::{
    model::AnnotationComponentType,
    update::{GraphUpdate, UpdateEvent},
};
use graphannis_core::graph::ANNIS_NS;
use itertools::Itertools;
use pest::{
    iterators::{Pair, Pairs},
    Parser,
};
use pest_derive::Parser;
use serde_derive::Deserialize;

use crate::{
    error::AnnattoError,
    util::graphupdate::import_corpus_graph_from_files,
    workflow::{StatusMessage, StatusSender},
    Module, StepID,
};

use super::Importer;

pub const MODULE_NAME: &str = "import_conllu";

#[derive(Default, Deserialize)]
#[serde(default)]
pub struct ImportCoNLLU {}

impl Module for ImportCoNLLU {
    fn module_name(&self) -> &str {
        MODULE_NAME
    }
}

impl Importer for ImportCoNLLU {
    fn import_corpus(
        &self,
        input_path: &std::path::Path,
        _step_id: StepID,
        tx: Option<crate::workflow::StatusSender>,
    ) -> Result<graphannis::update::GraphUpdate, Box<dyn std::error::Error>> {
        // TODO use ProgressReporter
        let mut update = GraphUpdate::default();
        let paths_and_node_names =
            import_corpus_graph_from_files(&mut update, input_path, &["conll", "conllu"])?;
        for (pathbuf, doc_node_name) in paths_and_node_names {
            if let Err(e) = self.import_document(&mut update, pathbuf.as_path(), doc_node_name, &tx)
            {
                if let Some(ref sender) = tx {
                    let reason = e.to_string();
                    sender.send(StatusMessage::Failed(AnnattoError::Import {
                        reason,
                        importer: self.module_name().to_string(),
                        path: input_path.to_path_buf(),
                    }))?;
                }
            }
        }
        Ok(update)
    }
}

impl ToString for Rule {
    fn to_string(&self) -> String {
        match self {
            Rule::lemma => "lemma".to_string(),
            Rule::upos => "upos".to_string(),
            Rule::xpos => "xpos".to_string(),
            Rule::deprel => "deprel".to_string(),
            Rule::enhanced_rel => "rel".to_string(),
            _ => "".to_string(),
        }
    }
}

type DepSpec = (usize, Option<String>);

impl ImportCoNLLU {
    fn import_document(
        &self,
        update: &mut GraphUpdate,
        document_path: &Path,
        document_node_name: String,
        tx: &Option<StatusSender>,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let f = std::fs::File::open(document_path)?;
        let mut decoder = DecodeReaderBytes::new(f);
        let mut file_content = String::new();
        decoder.read_to_string(&mut file_content)?; // TODO this needs to be buffered. UD Files can be very large
        let conllu: Pairs<Rule> = CoNLLUParser::parse(Rule::conll, &file_content)?;
        self.map_document(update, document_node_name, conllu, tx)?;
        Ok(())
    }

    fn map_document(
        &self,
        update: &mut GraphUpdate,
        document_node_name: String,
        mut conllu: Pairs<Rule>,
        tx: &Option<StatusSender>,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let mut token_names = Vec::new();
        if let Some(pair) = conllu.next() {
            if pair.as_rule() == Rule::conll {
                for sentence in pair.into_inner() {
                    // iterate over sentences
                    if sentence.as_rule() == Rule::sentence {
                        token_names.extend(self.map_sentence(
                            update,
                            document_node_name.as_str(),
                            sentence,
                            tx,
                        )?);
                    }
                }
            }
        } else if let Some(sender) = tx {
            let msg = format!("Could not parse file as conllu: {document_node_name}");
            let err = AnnattoError::Import {
                reason: msg,
                importer: self.module_name().to_string(),
                path: PathBuf::from(document_node_name),
            };
            sender.send(crate::workflow::StatusMessage::Failed(err))?;
        }
        for (source, target) in token_names.iter().tuple_windows() {
            update.add_event(UpdateEvent::AddEdge {
                source_node: source.to_string(),
                target_node: target.to_string(),
                layer: ANNIS_NS.to_string(),
                component_type: AnnotationComponentType::Ordering.to_string(),
                component_name: "".to_string(),
            })?;
        }
        Ok(())
    }

    fn map_sentence(
        &self,
        update: &mut GraphUpdate,
        document_node_name: &str,
        sentence: Pair<Rule>,
        tx: &Option<StatusSender>,
    ) -> Result<Vec<String>, Box<dyn std::error::Error>> {
        let mut id_to_tok_name = BTreeMap::new();
        let mut dependencies = Vec::new();
        let mut s_annos = Vec::new();
        let (l, c) = sentence.line_col();
        for member in sentence.into_inner() {
            match member.as_rule() {
                Rule::token => {
                    let (tok_name, tok_id, dep) =
                        self.map_token(update, document_node_name, member, tx)?;
                    id_to_tok_name.insert(tok_id, tok_name.to_string());
                    if let Some(dependency) = dep {
                        dependencies.push((tok_name, dependency.0, dependency.1));
                    }
                }
                Rule::multi_token | Rule::invalid_multi_token => {
                    // todo future work if needed
                }
                Rule::empty_node | Rule::invalid_empty_node => {
                    // todo future work if needed
                }
                Rule::s_anno => {
                    let mut name = None;
                    let mut value = None;
                    for name_or_s_value in member.into_inner() {
                        match name_or_s_value.as_rule() {
                            Rule::name => {
                                name = Some(name_or_s_value.as_str().trim().to_string());
                            }
                            Rule::s_value => {
                                value = Some(name_or_s_value.as_str().trim().to_string())
                            }
                            _ => {}
                        }
                        if let Some(ref n) = name {
                            if let Some(ref v) = value {
                                s_annos.push((n.to_string(), v.to_string()));
                                break;
                            }
                        }
                    }
                }
                _ => {}
            }
        }
        if !id_to_tok_name.is_empty() {
            let node_name = format!("{document_node_name}#s{l}_{c}");
            update.add_event(UpdateEvent::AddNode {
                node_name: node_name.to_string(),
                node_type: "node".to_string(),
            })?;
            for (anno_name, anno_value) in s_annos {
                update.add_event(UpdateEvent::AddNodeLabel {
                    node_name: node_name.to_string(),
                    anno_ns: "".to_string(),
                    anno_name: anno_name.to_string(),
                    anno_value: anno_value.to_string(),
                })?;
            }
            for token_name in id_to_tok_name.values() {
                update.add_event(UpdateEvent::AddEdge {
                    source_node: node_name.to_string(),
                    target_node: token_name.to_string(),
                    layer: ANNIS_NS.to_string(),
                    component_type: AnnotationComponentType::Coverage.to_string(),
                    component_name: "".to_string(),
                })?;
            }
            for (target_node_name, head_id, deprel) in dependencies {
                if head_id > 0 {
                    if let Some(source_node_name) = id_to_tok_name.get(&head_id) {
                        update.add_event(UpdateEvent::AddEdge {
                            source_node: source_node_name.to_string(),
                            target_node: target_node_name.to_string(),
                            layer: "".to_string(),
                            component_type: AnnotationComponentType::Pointing.to_string(),
                            component_name: "dep".to_string(),
                        })?;
                        if let Some(deprel_value) = deprel {
                            update.add_event(UpdateEvent::AddEdgeLabel {
                                source_node: source_node_name.to_string(),
                                target_node: target_node_name.to_string(),
                                layer: "".to_string(),
                                component_type: AnnotationComponentType::Pointing.to_string(),
                                component_name: "dep".to_string(),
                                anno_ns: "".to_string(),
                                anno_name: "deprel".to_string(),
                                anno_value: deprel_value.to_string(),
                            })?;
                        }
                    } else if let Some(sender) = tx {
                        let msg =
                            format!("Failed to build dependency tree: Unknown head id `{head_id}` ({l}, {c})");
                        let err = AnnattoError::Import {
                            reason: msg,
                            importer: self.module_name().to_string(),
                            path: Path::new(document_node_name).to_path_buf(),
                        };
                        sender.send(StatusMessage::Failed(err))?;
                    }
                }
            }
        }
        Ok(id_to_tok_name.into_iter().map(|e| e.1).collect_vec())
    }

    fn map_token(
        &self,
        update: &mut GraphUpdate,
        document_node_name: &str,
        token: Pair<Rule>,
        _tx: &Option<StatusSender>,
    ) -> Result<(String, usize, Option<DepSpec>), Box<dyn std::error::Error>> {
        let (l, c) = token.line_col();
        let line = token.as_str().to_string();
        let node_name = format!("{document_node_name}#t{l}_{c}");
        update.add_event(UpdateEvent::AddNode {
            node_name: node_name.to_string(),
            node_type: "node".to_string(),
        })?;
        update.add_event(UpdateEvent::AddNodeLabel {
            node_name: node_name.to_string(),
            anno_ns: ANNIS_NS.to_string(),
            anno_name: "layer".to_string(),
            anno_value: "default_layer".to_string(),
        })?;
        let mut token_id = None;
        let mut head_id = None;
        let mut deprel = None;
        for member in token.into_inner() {
            let rule = member.as_rule();
            match rule {
                Rule::id => {
                    token_id = Some(member.as_str().parse::<usize>()?);
                }
                Rule::form => {
                    update.add_event(UpdateEvent::AddNodeLabel {
                        node_name: node_name.to_string(),
                        anno_ns: ANNIS_NS.to_string(),
                        anno_name: "tok".to_string(),
                        anno_value: member.as_str().to_string(),
                    })?;
                }
                Rule::lemma | Rule::upos | Rule::xpos => {
                    let anno_name = rule.to_string();
                    update.add_event(UpdateEvent::AddNodeLabel {
                        node_name: node_name.to_string(),
                        anno_ns: "".to_string(),
                        anno_name: anno_name.to_string(),
                        anno_value: member.as_str().to_string(),
                    })?;
                }
                Rule::features | Rule::misc => {
                    for feature in member.into_inner() {
                        if feature.as_rule() == Rule::feature {
                            let mut anno_name = None;
                            let mut anno_value = None;
                            for name_or_value in feature.into_inner() {
                                let feature_rule = name_or_value.as_rule();
                                if feature_rule == Rule::name {
                                    anno_name = Some(name_or_value.as_str().to_string());
                                } else if feature_rule == Rule::value {
                                    anno_value = Some(name_or_value.as_str().to_string());
                                }
                                if anno_name.is_some() && anno_value.is_some() {
                                    update.add_event(UpdateEvent::AddNodeLabel {
                                        node_name: node_name.to_string(),
                                        anno_ns: "".to_string(),
                                        anno_name: anno_name.unwrap().trim().to_string(),
                                        anno_value: anno_value.unwrap().trim().to_string(),
                                    })?;
                                    break;
                                }
                            }
                        } else if feature.as_rule() == Rule::no_value {
                            break;
                        }
                    }
                }
                Rule::head => {
                    for id_or_else in member.into_inner() {
                        if id_or_else.as_rule() == Rule::id {
                            head_id = Some(id_or_else.as_str().trim().parse::<usize>()?);
                            break;
                        }
                    }
                }
                Rule::deprel => {
                    deprel = Some(member.as_str().trim().to_string());
                }
                Rule::enhanced_deps => {}
                _ => {}
            }
        }
        let dependency = head_id.map(|v| (v, deprel));
        if let Some(id) = token_id {
            Ok((node_name, id, dependency))
        } else {
            // by grammar spec this branch should never be possible
            let reason = format!("Token `{line}` ({l}, {c}) has no id which is invalid.");
            Err(Box::new(AnnattoError::Import {
                reason,
                importer: self.module_name().to_string(),
                path: document_node_name.into(),
            }))
        }
    }
}

#[derive(Parser)]
#[grammar = "importer/conllu/conllu.pest"]
struct CoNLLUParser;

#[cfg(test)]
mod tests;
