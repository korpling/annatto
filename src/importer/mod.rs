//! Contains importers and their traits.

pub mod graphml;

use crate::{workflow::StatusSender, Module, StepID};
use graphannis::update::GraphUpdate;
use std::{collections::BTreeMap, path::Path};

/// An importer is a module that takes a path and produces a list of graph update events.
/// Using the graph update event list allows to execute several importers in parallel and join them to a single annotation graph.
pub trait Importer: Module {
    /// Returns a list of graph update events for a single corpus.
    ///
    /// # Arguments
    ///
    /// * `input_path` - The path to the corpus files to import. Can be a single file or a directory. For directories, the importer should be able to find all relevant files in the directory.
    /// * `properties` - A map of configuration properties as given in the workflow description.
    /// * `tx` - If supported by the caller, this is a sender object that allows to send [status updates](../workflow/enum.StatusMessage.html) (like information messages, warnings and module progress) to the calling entity.
    ///
    fn import_corpus(
        &self,
        input_path: &Path,
        properties: &BTreeMap<String, String>,
        tx: Option<StatusSender>,
    ) -> Result<GraphUpdate, Box<dyn std::error::Error>>;
}

pub struct DoNothingImporter {}

impl DoNothingImporter {
    pub fn new() -> DoNothingImporter {
        DoNothingImporter {}
    }
}

impl Importer for DoNothingImporter {
    fn import_corpus(
        &self,
        path: &Path,
        _properties: &BTreeMap<String, String>,
        tx: Option<StatusSender>,
    ) -> Result<GraphUpdate, Box<dyn std::error::Error>> {
        if let Some(tx) = tx {
            let id = StepID {
                module_name: self.module_name().to_string(),
                path: Some(path.to_path_buf()),
            };
            tx.send(crate::workflow::StatusMessage::Progress { id, progress: 1.0 })?;
        }
        Ok(GraphUpdate::default())
    }
}

impl Module for DoNothingImporter {
    fn module_name(&self) -> &str {
        "DoNothingImporter"
    }
}
