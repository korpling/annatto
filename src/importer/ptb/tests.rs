use std::{collections::BTreeMap, io::BufWriter, path::PathBuf};

use graphannis::AnnotationGraph;
use insta::assert_snapshot;

use crate::importer::Importer;

use super::PtbImporter;

#[test]
fn ptb_oneline() {
    let properties: BTreeMap<String, String> = BTreeMap::new();

    let importer = PtbImporter::default();

    let mut u = importer
        .import_corpus(
            &PathBuf::from("tests/data/import/ptb/oneline"),
            &properties,
            None,
        )
        .unwrap();
    let mut g = AnnotationGraph::with_default_graphstorages(false).unwrap();
    g.apply_update(&mut u, |_| {}).unwrap();

    let mut buf = BufWriter::new(Vec::new());
    graphannis_core::graph::serialization::graphml::export(&g, None, &mut buf, |_| {}).unwrap();
    let bytes = buf.into_inner().unwrap();
    let actual = String::from_utf8(bytes).unwrap();

    assert_snapshot!(actual);
}
