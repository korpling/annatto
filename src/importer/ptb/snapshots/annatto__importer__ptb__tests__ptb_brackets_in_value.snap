---
source: src/importer/ptb/tests.rs
expression: actual
---
<?xml version="1.0" encoding="UTF-8"?>
<graphml>
    <key id="k0" for="node" attr.name="syntax::cat" attr.type="string"/>
    <key id="k1" for="node" attr.name="annis::doc" attr.type="string"/>
    <key id="k2" for="node" attr.name="annis::layer" attr.type="string"/>
    <key id="k3" for="node" attr.name="annis::node_type" attr.type="string"/>
    <key id="k4" for="node" attr.name="default_ns::pos" attr.type="string"/>
    <key id="k5" for="node" attr.name="annis::tok" attr.type="string"/>
    <key id="k6" for="node" attr.name="annis::tok-whitespace-before" attr.type="string"/>
    <graph edgedefault="directed" parse.order="nodesfirst" parse.nodeids="free" parse.edgeids="canonical">
        <node id="brackets_in_value">
            <data key="k3">corpus</data>
        </node>
        <node id="brackets_in_value/doc1">
            <data key="k3">corpus</data>
            <data key="k1">doc1</data>
        </node>
        <node id="brackets_in_value/doc1#text">
            <data key="k3">datasource</data>
        </node>
        <node id="brackets_in_value/doc1#n1">
            <data key="k3">node</data>
            <data key="k0">P</data>
            <data key="k2">syntax</data>
        </node>
        <node id="brackets_in_value/doc1#t1">
            <data key="k3">node</data>
            <data key="k2">default_layer</data>
            <data key="k5">(</data>
            <data key="k4">PUNCT</data>
        </node>
        <node id="brackets_in_value/doc1#t2">
            <data key="k3">node</data>
            <data key="k2">default_layer</data>
            <data key="k5">-RRB-</data>
            <data key="k4">NOUN</data>
            <data key="k6"> </data>
        </node>
        <node id="brackets_in_value/doc1#t3">
            <data key="k3">node</data>
            <data key="k2">default_layer</data>
            <data key="k5">)</data>
            <data key="k4">PUNCT</data>
            <data key="k6"> </data>
        </node>
        <edge id="e0" source="brackets_in_value/doc1#n1" target="brackets_in_value/doc1#t1" label="Dominance/syntax/">
        </edge>
        <edge id="e1" source="brackets_in_value/doc1#n1" target="brackets_in_value/doc1#t2" label="Dominance/syntax/">
        </edge>
        <edge id="e2" source="brackets_in_value/doc1#n1" target="brackets_in_value/doc1#t3" label="Dominance/syntax/">
        </edge>
        <edge id="e3" source="brackets_in_value/doc1#n1" target="brackets_in_value/doc1#t1" label="Dominance/syntax/edge">
        </edge>
        <edge id="e4" source="brackets_in_value/doc1#n1" target="brackets_in_value/doc1#t2" label="Dominance/syntax/edge">
        </edge>
        <edge id="e5" source="brackets_in_value/doc1#n1" target="brackets_in_value/doc1#t3" label="Dominance/syntax/edge">
        </edge>
        <edge id="e6" source="brackets_in_value/doc1#t2" target="brackets_in_value/doc1#t3" label="Ordering/annis/">
        </edge>
        <edge id="e7" source="brackets_in_value/doc1#t1" target="brackets_in_value/doc1#t2" label="Ordering/annis/">
        </edge>
        <edge id="e8" source="brackets_in_value/doc1#t2" target="brackets_in_value/doc1#text" label="PartOf/annis/">
        </edge>
        <edge id="e9" source="brackets_in_value/doc1#text" target="brackets_in_value/doc1" label="PartOf/annis/">
        </edge>
        <edge id="e10" source="brackets_in_value/doc1#t1" target="brackets_in_value/doc1#text" label="PartOf/annis/">
        </edge>
        <edge id="e11" source="brackets_in_value/doc1" target="brackets_in_value" label="PartOf/annis/">
        </edge>
        <edge id="e12" source="brackets_in_value/doc1#t3" target="brackets_in_value/doc1#text" label="PartOf/annis/">
        </edge>
    </graph>
</graphml>
